### PR TITLE
Refactor NuGet.Protocol v3 search: improve method names

### LINKED ISSUE #8478: Searching in PMUI with Static / Sleet package sources causes OutOfMemoryException
# Details about Problem

While using the PM UI and searching for a package, I received multiple 'System.OutOfMemoryException'. Discussing with @loic-sharma and @rrelyea, it seems we may need to discuss an approach to handling sleet / static sources in a better way.

Related past discussion: https://github.com/NuGet/Home/issues/7031
JSON Performance discussion: https://github.com/NuGetPackageExplorer/NuGetPackageExplorer/issues/804

NuGet product used: VS UI
VS version (if appropriate): 16.3.0 Preview 3.0

Worked before? If so, with which NuGet version: Related to new static sources.

## Detailed repro steps so we can see the same problem

1. Create a new ASP.NET Core Web Application.

2. Drop a new NuGet.Config file in the solution directory with the following sources:
```
<packageSources>
        <clear />
        <add key="dotnet-core" value="https://dotnetfeed.blob.core.windows.net/dotnet-core/index.json" />
        <add key="extensions" value="https://dotnetfeed.blob.core.windows.net/aspnet-extensions/index.json" />
        <add key="entityframeworkcore" value="https://dotnetfeed.blob.core.windows.net/aspnet-entityframeworkcore/index.json" />
        <add key="entityframework6" value="https://dotnetfeed.blob.core.windows.net/aspnet-entityframework6/index.json" />
        <add key="aspnetcore-tooling" value="https://dotnetfeed.blob.core.windows.net/aspnet-aspnetcore-tooling/index.json" />
        <add key="aspnetcore" value="https://dotnetfeed.blob.core.windows.net/aspnet-aspnetcore/index.json" />
        <add key="aspnet-blazor" value="https://dotnetfeed.blob.core.windows.net/aspnet-blazor/index.json" />
        <add key="NuGet.org" value="https://api.nuget.org/v3/index.json" />
    </packageSources>
  <disabledPackageSources>
    <clear />
  </disabledPackageSources>
```
3. Open the `Manage NuGet Packages` UI
4. Search for various packages. It should take 3-5 attempts to recreate the problem. Example: Microsoft.EntityFramework.X", where X is changed to various characters.
5. In the Output window, with "Show output from: Package Manager", you'll see exceptions similar to the following (could be 1 or more of these):
```[aspnetcore] Exception of type 'System.OutOfMemoryException' was thrown.
[dotnet-core] Exception of type 'System.OutOfMemoryException' was thrown.
[dotnet-core] Exception of type 'System.OutOfMemoryException' was thrown.
[dotnet-core] Exception of type 'System.OutOfMemoryException' was thrown.
[dotnet-core] Exception of type 'System.OutOfMemoryException' was thrown.
[aspnetcore] Exception of type 'System.OutOfMemoryException' was thrown.
```

### PR DESCRIPTION
**Pull request by erdembayar**
Bug
Fixes: NuGet/Home#8478

Fix
Details: Today I merged my [PR](https://github.com/NuGet/NuGet.Client/pull/3406) right before Nikolche left comments, so this PR is for addressing Nikolche code review. I refactored code according to it.

Testing/Validation
Tests Added: Modified existing unit test.
Reason for not adding tests:
Validation: Manual and unit test.

### CODE REVIEW

**Comment by erdembayar** [comment]
@nkolev92 @zivkan please review. I did changes we discussed this morning.

File: src/NuGet.Core/NuGet.Protocol/Resources/PackageSearchResourceV3.cs
Comment (nkolev92) [inline src/NuGet.Core/NuGet.Protocol/Resources/PackageSearchResourceV3.cs:16]: We need to keep the old constructor and make a switch in the code based on which constructor was used. 

That was consumers like @campersau are not broken in any way.

File: src/NuGet.Core/NuGet.Protocol/Resources/RawSearchResourceV3.cs
Comment (nkolev92) [inline src/NuGet.Core/NuGet.Protocol/Resources/RawSearchResourceV3.cs:16]: Did we mark this class as Obsolete?

Comment (erdembayar) [reply src/NuGet.Core/NuGet.Protocol/Resources/RawSearchResourceV3.cs:16]: No, because if do that then I have to make change to RawSearchResourceV3Provider or delete it. 
Is it ok to delete it RawSearchResourceV3Provider?
image

File: src/NuGet.Core/NuGet.Protocol/Resources/PackageSearchResourceV3.cs
Comment (erdembayar) [reply src/NuGet.Core/NuGet.Protocol/Resources/PackageSearchResourceV3.cs:16]: Ok. I added back old constructor along new one.

File: src/NuGet.Core/NuGet.Protocol/Providers/RawSearchResourceV3Provider.cs
Comment (erdembayar) [inline src/NuGet.Core/NuGet.Protocol/Providers/RawSearchResourceV3Provider.cs]: @nkolev92 I deleted your other comment for Obsolote attribute by mistake. We have Obsolote attribute at class level and method level for RawSearchResourceV3Provider.cs

File: src/NuGet.Core/NuGet.Protocol/Providers/PackageSearchResourceV3Provider.cs
Comment (erdembayar) [inline src/NuGet.Core/NuGet.Protocol/Providers/PackageSearchResourceV3Provider.cs:25]: @nkolev92 @zivkan 
Is it ok this one doesn't create new PackageSearchResourceV3(rawSearch) anymore even though we have constructor for PackageSearchResourceV3?

Comment (zivkan) [reply src/NuGet.Core/NuGet.Protocol/Providers/PackageSearchResourceV3Provider.cs:25]: as long as public APIs don't have a breaking change in behaviour, the implementation details don't matter.

File: src/NuGet.Core/NuGet.Protocol/Providers/RawSearchResourceV3Provider.cs
Comment (zivkan) [reply src/NuGet.Core/NuGet.Protocol/Providers/RawSearchResourceV3Provider.cs]: please provide a message so just in case someone is using this class, they know what to use instead.

File: src/NuGet.Core/NuGet.Protocol/Providers/PackageSearchResourceV3Provider.cs
Comment (nkolev92) [reply src/NuGet.Core/NuGet.Protocol/Providers/PackageSearchResourceV3Provider.cs:25]: The provider is about the default experience. 
And we're saying the default experience doesn't use the raw search resource and that's ok. 

We're keeping the constructor there for binary compatibility.

File: src/NuGet.Core/NuGet.Protocol/Providers/RawSearchResourceV3Provider.cs
Comment (zivkan) [inline src/NuGet.Core/NuGet.Protocol/Providers/RawSearchResourceV3Provider.cs]: I had to look up what CS0618 is. I don't support ignoring this warning thoughout the whole file. Given we want people to move to the new, better performance API, any API in this class that calls the old, low-performance API should ideally be modified to call the new, high-performance instead, and if that can't happen, then it should also be marked as obsolete. Repeat until there are no CS0618.

Only if this starts to affect a lot of files, we'll reconsider, but we should be trying to the "right thing" first, and hide warnings as a last resort.

File: src/NuGet.Core/NuGet.Protocol/Resources/PackageSearchResourceV3.cs
Comment (zivkan) [inline src/NuGet.Core/NuGet.Protocol/Resources/PackageSearchResourceV3.cs]: Mark the constructor as obsolete, rather than disabling the warning.

File: src/NuGet.Core/NuGet.Protocol/Providers/PackageSearchResourceV3Provider.cs
Comment (nkolev92) [reply src/NuGet.Core/NuGet.Protocol/Providers/PackageSearchResourceV3Provider.cs:25]: The only contract that the provider has is that it returns a PackageSearchResource. 

So it's ok that the RawSearchResource is not used.

File: src/NuGet.Core/NuGet.Protocol/Providers/RawSearchResourceV3Provider.cs
Comment (nkolev92) [reply src/NuGet.Core/NuGet.Protocol/Providers/RawSearchResourceV3Provider.cs]: lgtm

File: src/NuGet.Core/NuGet.Protocol/Resources/PackageSearchResourceV3.cs
Comment (nkolev92) [inline src/NuGet.Core/NuGet.Protocol/Resources/PackageSearchResourceV3.cs:25]: We should use global supressions instead.

File: src/NuGet.Core/NuGet.Protocol/Repository.cs
Comment (nkolev92) [inline src/NuGet.Core/NuGet.Protocol/Repository.cs]: Use GlobalSupressions instead.

File: src/NuGet.Core/NuGet.Protocol/Resources/PackageSearchResourceV3.cs
Comment (nkolev92) [inline src/NuGet.Core/NuGet.Protocol/Resources/PackageSearchResourceV3.cs]: you can probably simplify this with coalesce operator

File: src/NuGet.Core/NuGet.Protocol/Resources/PackageSearchResourceV3.cs
Comment (zivkan) [inline src/NuGet.Core/NuGet.Protocol/Resources/PackageSearchResourceV3.cs]: do we need to make this constructor public?  People should be using `PackageSearchResource`, not this class directly.

File: src/NuGet.Core/NuGet.Protocol/Resources/PackageSearchResourceV3.cs
Comment (nkolev92) [inline src/NuGet.Core/NuGet.Protocol/Resources/PackageSearchResourceV3.cs]: Same here, use GlobalSuppressions.

Comment (zivkan) [inline src/NuGet.Core/NuGet.Protocol/Resources/PackageSearchResourceV3.cs]: Why `618`, and not `CS0618`? Higher in the same file, the longer format was used, which I think is more clear.

File: test/NuGet.Core.Tests/NuGet.Protocol.Tests/RawSearchResourceTests.cs
Comment (nkolev92) [inline test/NuGet.Core.Tests/NuGet.Protocol.Tests/RawSearchResourceTests.cs]: I wouldn't delete the whole class. 

There were some old tests unrelated to your work such as `RawSearchResource_SearchEncoding` for example (the 1st test).

**Review by nkolev92 — COMMENTED** [review]
Looks great. A couple of small comments.

File: src/NuGet.Core/NuGet.Protocol/Providers/PackageSearchResourceV3Provider.cs
Comment (zivkan) [inline src/NuGet.Core/NuGet.Protocol/Providers/PackageSearchResourceV3Provider.cs]: Doesn't the current code throw `FatalProtocolException(Strings.Protocol_MissingSearchService)` in this scenario?  Your change would not, and would return a tuple (false, null). This looks like a change in behaviour, unless I missed the existing code that does the `> 0` check.

File: src/NuGet.Core/NuGet.Protocol/Resources/PackageSearchResourceV3.cs
Comment (zivkan) [inline src/NuGet.Core/NuGet.Protocol/Resources/PackageSearchResourceV3.cs]:
```suggestion
                throw new ArgumentNullException(nameof(searchEndpoints));
```

this should allow you to remove one of the new lines in the global suppression file.

File: src/NuGet.Core/NuGet.Protocol/GlobalSuppressions.cs
Comment (zivkan) [inline src/NuGet.Core/NuGet.Protocol/GlobalSuppressions.cs]: sorry, but the justification doesn't make sense for the warning being suppressed.  I provided a suggestion in the code which should remove the need for this (line 304) suppression.

File: src/NuGet.Core/NuGet.Protocol/Providers/RawSearchResourceV3Provider.cs
Comment (nkolev92) [reply src/NuGet.Core/NuGet.Protocol/Providers/RawSearchResourceV3Provider.cs]: I wasn't suggesting that. GlobalSuppressions allows you to suppress warnings for specific lines only.

File: src/NuGet.Core/NuGet.Protocol/Repository.cs
Comment (erdembayar) [reply src/NuGet.Core/NuGet.Protocol/Repository.cs]: Problem was VS was not offering any option to suppress this into GlobalSuppressions file. It seems other soIution is have to suppress for whole GetCoreV3() method or this file.

File: src/NuGet.Core/NuGet.Protocol/Resources/PackageSearchResourceV3.cs
Comment (erdembayar) [reply src/NuGet.Core/NuGet.Protocol/Resources/PackageSearchResourceV3.cs:25]: Same here VS was not offering any option to suppress this into GlobalSuppressions file. It seems other soIution is have to suppress for whole this file.

File: src/NuGet.Core/NuGet.Protocol/Resources/PackageSearchResourceV3.cs
Comment (erdembayar) [reply src/NuGet.Core/NuGet.Protocol/Resources/PackageSearchResourceV3.cs]: Ok. Made it internal.

File: src/NuGet.Core/NuGet.Protocol/GlobalSuppressions.cs
Comment (erdembayar) [reply src/NuGet.Core/NuGet.Protocol/GlobalSuppressions.cs]: Ok. Fixed.

File: src/NuGet.Core/NuGet.Protocol/Providers/PackageSearchResourceV3Provider.cs
Comment (erdembayar) [reply src/NuGet.Core/NuGet.Protocol/Providers/PackageSearchResourceV3Provider.cs]: Added throw exception on else condition, I got this logic from RawSearchResourceProvider, which didn't have any exception.
Can you check my change now?

File: src/NuGet.Core/NuGet.Protocol/Providers/RawSearchResourceV3Provider.cs
Comment (erdembayar) [reply src/NuGet.Core/NuGet.Protocol/Providers/RawSearchResourceV3Provider.cs]: Removed it, I inserted wrong place.

File: src/NuGet.Core/NuGet.Protocol/Resources/PackageSearchResourceV3.cs
Comment (erdembayar) [reply src/NuGet.Core/NuGet.Protocol/Resources/PackageSearchResourceV3.cs]: Ok. Done.

File: src/NuGet.Core/NuGet.Protocol/Resources/PackageSearchResourceV3.cs
Comment (erdembayar) [reply src/NuGet.Core/NuGet.Protocol/Resources/PackageSearchResourceV3.cs]: Done.

File: src/NuGet.Core/NuGet.Protocol/Resources/PackageSearchResourceV3.cs
Comment (erdembayar) [reply src/NuGet.Core/NuGet.Protocol/Resources/PackageSearchResourceV3.cs]: Same here VS was not offering any option to suppress this into GlobalSuppressions file. It seems other soIution is have to suppress for whole this file.

Comment (erdembayar) [reply src/NuGet.Core/NuGet.Protocol/Resources/PackageSearchResourceV3.cs]: Aren't they same thing?

File: src/NuGet.Core/NuGet.Protocol/Providers/RawSearchResourceV3Provider.cs
Comment (erdembayar) [reply src/NuGet.Core/NuGet.Protocol/Providers/RawSearchResourceV3Provider.cs]: Added.

File: test/NuGet.Core.Tests/NuGet.Protocol.Tests/RawSearchResourceTests.cs
Comment (erdembayar) [reply test/NuGet.Core.Tests/NuGet.Protocol.Tests/RawSearchResourceTests.cs]: Exact same tests with PackageSearchResourceV3 implementation moved into PackageSearchResourceV3Tests.cs , so they're still covered.
I felt we're duplicate tests.

File: src/NuGet.Core/NuGet.Protocol/Resources/PackageSearchResourceV3.cs
Comment (srdjanjovcic) [reply src/NuGet.Core/NuGet.Protocol/Resources/PackageSearchResourceV3.cs:25]: You cannot suppress compiler warnings using attributes.

File: src/NuGet.Core/NuGet.Protocol/Resources/PackageSearchResourceV3.cs
Comment (srdjanjovcic) [reply src/NuGet.Core/NuGet.Protocol/Resources/PackageSearchResourceV3.cs]: This is CS0618, compiler warning, and cannot be suppressed using attributes. They have to be suppressed using `#pragma`s

File: src/NuGet.Core/NuGet.Protocol/Providers/PackageSearchResourceV3Provider.cs
Comment (zivkan) [reply src/NuGet.Core/NuGet.Protocol/Providers/PackageSearchResourceV3Provider.cs]: now the "business logic" of throwing this exception in this scenario is in two places. Is there are reason you don't remove the `if` check, and let `PackageSearchResourceV3` throw the exception, just as it does with the current production code?

File: src/NuGet.Core/NuGet.Protocol/Resources/PackageSearchResourceV3.cs
Comment (zivkan) [reply src/NuGet.Core/NuGet.Protocol/Resources/PackageSearchResourceV3.cs]: yes they are. Why aren't we consistent with which syntax we use?

File: src/NuGet.Core/NuGet.Protocol/Resources/PackageSearchResourceV3.cs
Comment (zivkan) [inline src/NuGet.Core/NuGet.Protocol/Resources/PackageSearchResourceV3.cs]: nitpick: extra whitespace line (see [style guide 7](https://github.com/NuGet/NuGet.Client/blob/dev/docs/coding-guidelines.md))

File: test/NuGet.Core.Tests/NuGet.Protocol.Tests/RawSearchResourceTests.cs
Comment (zivkan) [reply test/NuGet.Core.Tests/NuGet.Protocol.Tests/RawSearchResourceTests.cs]: They test different classes. Even if we mark `RawSearchResource` as obsolete, it's still part of our public API and should be supported as well as possible, including minimising the risk of regressions.

If we don't want to duplicate test code, we should have the test setup and test assertion code in one place, and execute it twice, once with `RawSearchResource` and once with `PackageSearchResourceV3`. In fact, doing it this way minimises risk of a future change changing behaviour in one of the two implementations.

File: src/NuGet.Core/NuGet.Protocol/Providers/PackageSearchResourceV3Provider.cs
Comment (nkolev92) [reply src/NuGet.Core/NuGet.Protocol/Providers/PackageSearchResourceV3Provider.cs]: I'm with @zivkan 

The PackageSearchResourceV3 should throw if it gets an empty collection of endpoints.

File: src/NuGet.Core/NuGet.Protocol/Resources/PackageSearchResourceV3.cs
Comment (nkolev92) [reply src/NuGet.Core/NuGet.Protocol/Resources/PackageSearchResourceV3.cs:25]: Resolved offline

File: src/NuGet.Core/NuGet.Protocol/Resources/PackageSearchResourceV3.cs
Comment (nkolev92) [inline src/NuGet.Core/NuGet.Protocol/Resources/PackageSearchResourceV3.cs:34]: This feels like it should be public 

@zivkan thoughts?

File: test/NuGet.Core.Tests/NuGet.Protocol.Tests/RawSearchResourceTests.cs
Comment (nkolev92) [reply test/NuGet.Core.Tests/NuGet.Protocol.Tests/RawSearchResourceTests.cs]: Yep, we should ensure our public APIs are tested even if they are unlikely to receive future updates.

File: src/NuGet.Core/NuGet.Protocol/Resources/PackageSearchResourceV3.cs
Comment (erdembayar) [reply src/NuGet.Core/NuGet.Protocol/Resources/PackageSearchResourceV3.cs:34]: Actually @zivkan recommended internal.

File: src/NuGet.Core/NuGet.Protocol/Providers/PackageSearchResourceV3Provider.cs
Comment (erdembayar) [reply src/NuGet.Core/NuGet.Protocol/Providers/PackageSearchResourceV3Provider.cs]: Ok. Removed condition so let PackageSearchResourceV3 handle it.

File: src/NuGet.Core/NuGet.Protocol/Resources/PackageSearchResourceV3.cs
Comment (erdembayar) [reply src/NuGet.Core/NuGet.Protocol/Resources/PackageSearchResourceV3.cs]: Ok, made it consistent. Somehow by default VS putting just 618 instead of CS0618.

File: src/NuGet.Core/NuGet.Protocol/Resources/PackageSearchResourceV3.cs
Comment (erdembayar) [reply src/NuGet.Core/NuGet.Protocol/Resources/PackageSearchResourceV3.cs]: Removed.

File: test/NuGet.Core.Tests/NuGet.Protocol.Tests/RawSearchResourceTests.cs
Comment (erdembayar) [reply test/NuGet.Core.Tests/NuGet.Protocol.Tests/RawSearchResourceTests.cs]: Ok. Restored it.

File: src/NuGet.Core/NuGet.Protocol/Resources/PackageSearchResourceV3.cs
Comment (nkolev92) [reply src/NuGet.Core/NuGet.Protocol/Resources/PackageSearchResourceV3.cs:34]: I see. 
Let's leave it then.

**Review by nkolev92 — APPROVED** [review]
I think it looks good. 

Wait on a 2nd sign-off.

**Review by zivkan — APPROVED** [review]

**Pull request merged**
